pdbtools/pdb_addter.py

```diff
@@ -0,0 +1,247 @@
+    if len(slices) == 3:
+        start, end, step = slices
+    elif len(slices) == 2:
+        start, end = slices
+    elif len(slices) == 1:
+        if option.startswith(':'):
+            end = slices[0]
+        elif option.endswith(':'):
+            start = slices[0]
+
+    if start is None:
+        emsg = 'ERROR!! Please specify starting value: \'{}\'\n'
+        sys.stderr.write(emsg.format(start))
+        sys.exit(1)
+    elif not is_integer(start):
+        emsg = 'ERROR!! Starting value must be numerical: \'{}\'\n'
+        sys.stderr.write(emsg.format(start))
+        sys.exit(1)
+    elif not (-999 <= int(start) < 9999):
+        emsg = 'ERROR!! Starting value must be between -999 and 9998: \'{}\'\n'
+        sys.stderr.write(emsg.format(start))
+        sys.exit(1)
+
+    if end is None:
+        end = 10000  # residues cannot reach this value (max 4 char)
+    elif not is_integer(end):
+        emsg = 'ERROR!! End value must be numerical: \'{}\'\n'
+        sys.stderr.write(emsg.format(end))
+        sys.exit(1)
+    elif not (-999 <= int(end) < 9999):
+        emsg = 'ERROR!! End value must be between -999 and 9998: \'{}\'\n'
+        sys.stderr.write(emsg.format(end))
+        sys.exit(1)
+
+    if step is None:
+        step = 1
+    elif not is_integer(step):
+        emsg = 'ERROR!! Step value must be numerical: \'{}\'\n'
+        sys.stderr.write(emsg.format(step))
+        sys.exit(1)
+    elif int(step) <= 0:
+        emsg = 'ERROR!! Step value must be a positive number: \'{}\'\n'
+        sys.stderr.write(emsg.format(step))
+        sys.exit(1)
+
+    start, end, step = int(start), int(end), int(step)
+
+    if start >= end:
+        emsg = 'ERROR!! Start ({}) cannot be larger than end ({})\n'
+        sys.stderr.write(emsg.format(start, end))
+        sys.exit(1)
+
+    resrange = set(range(start, end + 1))
+    return (fh, resrange, step)
+
+
+def run(fhandle, residue_range, step):
+    """
+    Add TER records within the residue range at frequency identified by step.
+
+    This function is a generator.
+
+    Parameters
+    ----------
+    fhandle : a line-by-line iterator of the original PDB file.
+
+    residue_range : set, list, or tuple
+        The residues describing the range to consider.
+
+    step : int
+        The step at which to insert a TER record.
+
+    Yields
+    ------
+    str (line-by-line)
+        All lines with added TER lines designated by inputs.
+    """
+    prev_res = None
+    res_counter = 0
+    records = ('ATOM')
+    for line in fhandle:
+        if line.startswith(records):
+
+            res_id = line[21:26]  # include chain ID
+            if res_id != prev_res:
+
+                prev_res = res_id
+                res_counter += 1
+                if res_counter - min(residue_range) != 0 and (res_counter - min(residue_range)) % step == 0 and int(line[22:26]) in residue_range:
+                    yield "TER\n"
+
+        yield line
```

Comment:
If `python pdbtools/pdb_addter.py -6:7:1 tests/data/dummy.pdb`, I get:

```
ATOM      1  N   ARG B   4      36.898  42.175  -2.688  1.00  0.00           N  
ATOM      2  H   ARG B   4      37.673  41.570  -2.916  1.00  0.00           H  
ATOM      3  H2  ARG B   4      35.929  41.470  -2.758  1.00  0.00           H  
ATOM      3  H3  ARG B   4      37.099  42.392  -1.524  1.00  0.00           H  
ATOM      3  CA  ARG B   4      37.080  43.455  -3.421  1.00  0.00           C  
ATOM      3  HA  ARG B   4      38.102  43.960  -3.065  1.00  0.00           H  
ATOM      3  CB  ARG B   4      37.064  43.172  -4.926  1.00  0.00           C  
ATOM      3  HB2 ARG B   4      36.937  44.238  -5.450  1.00  0.00           H  
ATOM      3  HB3 ARG B   4      36.146  42.550  -5.381  1.00  0.00           H  
ATOM      3  CG  ARG B   4      38.366  42.367  -5.167  1.00  0.00           C  
ATOM      3  HG2 ARG B   4      38.044  41.263  -5.501  1.00  0.00           H  
ATOM      3  HG3 ARG B   4      39.214  42.156  -4.350  1.00  0.00           H  
ATOM      3  CD  ARG B   4      39.191  43.008  -6.234  1.00  0.00           C  
ATOM      3  HD2 ARG B   4      39.652  44.053  -5.873  1.00  0.00           H  
ATOM      3  HD3 ARG B   4      40.211  42.425  -6.479  1.00  0.00           H  
ATOM      3  NE  ARG B   4      38.557  43.081  -7.486  1.00  0.00           N  
ATOM      3  HE  ARG B   4      38.421  41.980  -7.920  1.00  0.00           H  
ATOM      3  CZ  ARG B   4      38.266  43.951  -8.427  1.00  0.00           C  
ATOM      3  NH1 ARG B   4      38.554  45.234  -8.396  1.00  0.00           N  
ATOM      3 HH11 ARG B   4      39.085  45.950  -7.607  1.00  0.00           H  
ATOM      3 HH12 ARG B   4      38.392  45.917  -9.359  1.00  0.00           H  
ATOM      3  NH2 ARG B   4      37.609  43.458  -9.493  1.00  0.00           N  
ATOM      3 HH21 ARG B   4      36.480  43.750  -9.739  1.00  0.00           H  
ATOM      3 HH22 ARG B   4      38.086  42.813 -10.373  1.00  0.00           H  
ATOM      3  C   ARG B   4      36.102  44.524  -2.998  1.00  0.00           C  
ATOM      3  O   ARG B   4      36.577  45.677  -2.879  1.00  0.00           O  
TER
ATOM      3  N   GLU B   6      34.849  44.167  -2.710  1.00  0.00           N  
ATOM      3  H   GLU B   6      34.444  43.155  -3.175  1.00  0.00           H  
ATOM      3  CA  GLU B   6      33.861  45.127  -2.233  1.00  0.00           C  
ATOM      3  HA  GLU B   6      33.871  46.090  -2.939  1.00  0.00           H  
ATOM      3  CB  GLU B   6      32.480  44.538  -2.041  1.00  0.00           C  
ATOM      3  HB2 GLU B   6      32.305  43.781  -1.133  1.00  0.00           H  
ATOM      3  HB3 GLU B   6      31.793  45.504  -1.851  1.00  0.00           H  
ATOM      3  CG  GLU B   6      31.968  43.831  -3.261  1.00  0.00           C  
ATOM      3  HG2 GLU B   6      32.228  42.682  -3.461  1.00  0.00           H  
ATOM      3  HG3 GLU B   6      32.059  44.448  -4.281  1.00  0.00           H  
ATOM      3  CD  GLU B   6      30.440  43.676  -3.190  1.00  0.00           C  
ATOM      3  OE1 GLU B   6      29.816  44.208  -2.210  1.00  0.00           O  
ATOM      3  OE2 GLU B   6      30.005  43.060  -4.204  1.00  0.00           O  
ATOM      3  C   GLU B   6      34.180  45.629  -0.820  1.00  0.00           C  
ATOM      3  O   GLU B   6      33.914  46.775  -0.464  1.00  0.00           O  
TER
ATOM      3  N   ALA B   7      34.725  44.679  -0.087  1.00  0.00           N  
ATOM      3  H   ALA B   7      34.813  43.714  -0.371  1.00  0.00           H  
ATOM      3  CA  ALA B   7      35.081  45.036   1.305  1.00  0.00           C  
ATOM      3  HA  ALA B   7      34.200  45.604   1.876  1.00  0.00           H  
ATOM      3  CB  ALA B   7      35.381  43.810   2.140  1.00  0.00           C  
ATOM      3  HB1 ALA B   7      34.382  43.229   2.456  1.00  0.00           H  
ATOM      3  HB2 ALA B   7      36.167  42.973   1.802  1.00  0.00           H  
ATOM      3  HB3 ALA B   7      35.847  44.180   3.181  1.00  0.00           H  
ATOM      3  C   ALA B   7      36.213  46.067   1.258  1.00  0.00           C  
ATOM      3  O   ALA B   7      36.287  47.046   2.028  1.00  0.00           O  
TER       3      ALA B   7
```

Is the `TER` after `4` expected? Does it add it because `6` was first so it adds `TER` just before the start?

Reply:
I think this is because of using `int(line[22:26]) in residue_range` to partly  determine if it's the correct place of place ter when I should be using `res_counter`.

Reply:
think that python is 0-indexed. However, PDB files are not (though there can be negative residue numbers even) but the concept is not 0-indexed. Also, a user using `pdb-tools` is not 0-index minded. think about that when developing the tool.